packages/amplify_api_rest/test/amplify_api_rest_methods_test.dart

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import 'dart:typed_data';
+
+import 'package:flutter/services.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:amplify_api_rest/amplify_api_rest.dart';
+
+void main() {
```

Comment:
Need tests for error handling and cancel functionality.

Reply:
How would you recommend unit testing the cancel functionality if it's a void return type and just reacts directly to what the method channel returns?  We have unit tests for cancel functionality added for native. @Amplifiyer 

```
  @override
  void cancelRequest(String code) async {
    print("Attempting to cancel RestOperation " + code);

    await _channel.invokeMethod("cancel", code).then((result) {
      print("Cancel succeeded for RestOperation: " + code);
    }).catchError((e) {
      print("Cancel request failed due to: " + e.message + " " + e.code);
    });
  }
```

Reply:
You can test by mocking the response (behavior) of `_channel.invokeMethod("cancel", code)`. For one case when it doesn't throw an exception and in other when it does.

Reply:
Hmm perhaps I'm not seeing it correctly but it seems like that test would be quite basic in the sense that we're pretty much directly giving the answer to our code?

So you're saying to make _channel.invokeMethod("cancel", code) to not throw an exception and then to check that an exception is not thrown. 

And then to make _channel.invokeMethod("cancel", code) thrown an exception and verify that that same exception is thrown? 

@Amplifiyer

Reply:
Correct, I'm guessing you will be wrapping up the error and throwing a different exception than the `PlatformException` which can be tested here. Furthermore, we should test the following code if not already tested, specifically that the code generated is the one passed to the cancelRequest, basically a higher level method test for `cancelRequest`. Mae a `get` call -> called `cancelRequest` before and after the response future is resolved.

```dart
    try {
      // Send Request Code to Native
      String code = UUID.getUUID();

      Future<RestResponse> futureResponse =
          _callNativeRestMethod(methodName, code, restOptions);

      RestOperation restOperation = new RestOperation(
          response: futureResponse, cancel: () => {cancelRequest(code)});

      return restOperation;
    } on PlatformException catch (e) {
      throw (e);
    }
```

Reply:
No we don't throw any error when calling cancelRequest.  This is to be consistent with native which does not provide a result after calling cancel.  Instead the return type is cancel and no exception is thrown if you call it multiple times.  

Doing the test you suggested above would require renaming all of those methods to be public instead of private thereby exposing our internal functionality to users for the purpose of making that unit test.  I don't think that unit test is quite necessary as we directly pass the token generated to the cancelRequest method in our code already: 

```
    String cancelToken = UUID.getUUID();

    Future<RestResponse> futureResponse =
        _callNativeRestMethod(methodName, cancelToken, restOptions);

    return new RestOperation(
        response: futureResponse, cancel: () => {cancelRequest(cancelToken)});
```

Reply:
>Doing the test you suggested above would require renaming all of those methods to be public instead of private thereby exposing our internal functionality to users for the purpose of making that unit test. 

you can make them package protected for using in the unit tests which won't open them up for public.